### DISCUSSION TITLE
Using secrets for code and database credentials

### DISCUSSION THREAD

**Discussion by HSoeren** [community/community#26264] [Actions]
<p>Hi there</p>
<p>I am currently working with colleagues on a small webapp that is based on PHP and a database.</p>
<p>We have set up a workflow where every push to the master branch copies the repository to our test webserver.</p>
<p>Until now we have always worked with ZIP files and email - now that is automated and simplified but now we fail because the secrets do not work as we want them to.</p>
<p>We don’t want to store the real database accesses, but rather a variable that is replaced by the real credential when it is delivered to the web server, so that the web server gets a running webapp while there is no critical data in the repository.</p>
<p>Is there a way to automate this or do we have to upload manually again?</p>
<p>Thank you very much in advance!</p>

**Comment by Simran-B** [comment]
<p>That’s what GitHub secrets are, aren’t they?</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets" target="_blank" rel="noopener nofollow ugc">Encrypted secrets - GitHub Docs</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by HSoeren** [comment]
<p>Yes exactly. I use them always for the Action / Deploy process to store my FTP credentials in it. But now i have a repository cotaining a php file with credentials for my database and i dont want to expose them. I hoped that there is a way to hide them with with secrets.</p>

**Comment by HSoeren** [comment]
<p>idea; can I create a file via <code>cat ${{ SECRET }} &gt; file.php</code> in the workflow? If so, I would have the file containing the access data created by the workflow…?</p>

**Comment by airtower-luna** [comment]
<p>I haven’t tested it personally, but the documentation describes <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets#limits-for-secrets">how to commit an encrypted file and decrypt it using a secret</a>. Maybe that would help?</p>

**Comment by Simran-B** [comment]
<p>Having any sort of credentials hardcoded in a PHP file sounds like a bad idea. Why don’t you you let PHP read the credentials from environment variables for instance?</p>

**Comment by HSoeren** [comment]
<p>Can you be more specific? I know there is such a thing, but I don’t think I’ve really worked with it.</p>

**Comment by Simran-B** [comment, marked as answer]
<p>Something along the following, with the database password stored as GitHub secret with name <code>SCRIPT_CREDENTIALS</code>:</p>
<pre><code>steps:
  - name: Execute script
    env:
      PASSWORD: ${{ secrets.SCRIPT_CREDENTIALS }}
    run: php script.php
</code></pre>
<p><em>script.php</em>:</p>
<pre><code class="lang-auto">&lt;?php
$servername = "localhost";
$username = "username";
$password = getenv("PASSWORD");

$conn = new mysqli($servername, $username, $password);

if ($conn-&gt;connect_error) {
  die("Connection failed: " . $conn-&gt;connect_error);
}
echo "Connected successfully";
?&gt;
</code></pre>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://www.php.net/favicon.ico" class="site-icon" width="16" height="16">
      <a href="https://www.php.net/manual/en/function.getenv.php" target="_blank" rel="noopener nofollow ugc">php.net</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="16" height="16">

<h3><a href="https://www.php.net/manual/en/function.getenv.php" target="_blank" rel="noopener nofollow ugc">PHP: getenv - Manual</a></h3>



  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by HSoeren** [comment]
<p>Sorry, I was on the road for a few days and could not test nor answer anything.</p>
<p>Your solution looks good, would do that.</p>
<p>In the converted file on the server you will only find the variables but not the modified data (the content of the variables)</p>
<pre><code class="lang-auto">Run php config/db.php
  php config/db.php
  shell: /bin/bash -e {0}
  env:
    SERVER: ***
    USERNAME: ***
    PASSWORD: ***
PHP Warning:  mysqli::__construct(): (HY000/2002): No such file or directory in /home/runner/work/dcd-test/dcd-test/config/db.php on line 6
PHP Stack trace:
PHP   1. {main}() /home/runner/work/dcd-test/dcd-test/config/db.php:0
PHP   2. mysqli-&gt;__construct() /home/runner/work/dcd-test/dcd-test/config/db.php:6
Connection failed: No such file or directory
</code></pre>
<p>The folder is in the correct path on the server, it just seems like the php command doesn’t work. Does that need a special package?</p>

**Comment by Simran-B** [comment]
<p>Can you check what the working directory is with <code>pwd</code>?</p>
<p>Looks like it cannot find <code>config/db.php</code> relative to the current working directory. Although the last log entry says “Connection failed”, which seems a bit odd.</p>

**Comment by canovasjm** [comment]
<p>Thanks a lot for this!. I want to do the same using <a href="https://stackoverflow.com/questions/65315121/read-email-credentials-stored-as-gh-secret" rel="noopener nofollow ugc">python and email credentials</a>, instead. In python I’m using <code>os.environ.get()</code> to read the secrets, but with no success so far. Any ideas on how to do it with python?</p>

**Comment by airtower-luna** [comment]
<p>To quote from your Stackoverflow post:</p>
<pre><code class="lang-yaml">- name: execute py script # run sj-gobierno.py to get the latest data
  env: | 
    EMAIL_ADDRESS: ${{ secrets.EMAIL_ADDRESS }}
    EMAIL_PASSWORD: ${{ secrets.EMAIL_PASSWORD }}
</code></pre>
<p>The problem is that you’re turning setting a text block for <code>env</code>, instead of individual keys. Correct version:</p>
<pre><code class="lang-yaml">- name: execute py script # run sj-gobierno.py to get the latest data
  env:
    EMAIL_ADDRESS: ${{ secrets.EMAIL_ADDRESS }}
    EMAIL_PASSWORD: ${{ secrets.EMAIL_PASSWORD }}
</code></pre>
<p>I also highly recommend moving the pip install to a separate, previous step so it doesn’t have access to the secrets. Principle of least privilege and all that.</p>

**Comment by canovasjm** [comment]
<p>Thank you, it worked!. And thank you for the recommendation to move <code>pip install</code> to a previous step. For others interested in using secrets inside python, I can confirm that <code>os.environ.get()</code> does the job.</p>

**Comment by LaxmanMaharjan** [comment]
<p>.py file</p>
<div class="md-table">
<table>
<thead>
<tr>
<th></th>
<th>import smtplib</th>
</tr>
</thead>
<tbody>
<tr>
<td></td>
<td>import os</td>
</tr>
<tr>
<td></td>
<td>from email.message import EmailMessage</td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>EMAIL_ADDRESS = os.environ[‘EMAIL_ADDRESS’]</td>
</tr>
<tr>
<td></td>
<td>EMAIL_PASSWORD = os.environ[‘EMAIL_PASSWORD’]</td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>msg = EmailMessage()</td>
</tr>
<tr>
<td></td>
<td>msg[‘Subject’] = ‘Checking smtplib FROM GITHUB’</td>
</tr>
<tr>
<td></td>
<td>msg[‘From’] = EMAIL_ADDRESS</td>
</tr>
<tr>
<td></td>
<td>msg[‘To’] = ‘lxmnmrzn@gmail.com’</td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>msg.set_content(‘This is a email check.’)</td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td></td>
</tr>
<tr>
<td></td>
<td>with smtplib.SMTP_SSL(‘<a href="http://smtp.gmail.com" rel="noopener nofollow ugc">smtp.gmail.com</a>’, 465) as smtp:</td>
</tr>
<tr>
<td></td>
<td>smtp.login(EMAIL_ADDRESS, EMAIL_PASSWORD)</td>
</tr>
<tr>
<td></td>
<td>smtp.send_message(msg)</td>
</tr>
</tbody>
</table>
</div><p>.yml file</p>
<ul>
<li>name: execute py script<br>
env:<br>
EMAIL_ADDRESS: <span class="math">{{ secrets.EMAIL_ADDRESS }}
  EMAIL_PASSWORD: </span>{{ secrets.EMAIL_PASSWORD }}</li>
</ul>
<p>ERROR<br>
smtp.login(EMAIL_ADDRESS, EMAIL_PASSWORD)<br>
can only concatenate str (not “bytes”) to str</p>
<p>Please help me solve this problem</p>

**Comment by Simran-B** [comment]
<p>Are you using Python 2?<br>
<code>os.environ['...'].decode('utf8')</code> might do the trick.</p>

**Comment by LaxmanMaharjan** [comment]
<p>There was an issue with the GitHub secrets, was setting secrets in environment instead of repo.<br>
<a class="mention" href="/canovasjm">@canovasjm</a> helped me solve it.<br>
Thank you <a class="mention" href="/simran-b">@Simran-B</a> <a class="mention" href="/canovasjm">@canovasjm</a></p>

**Comment by maheshj01** [comment]
<p>I am trying to use the above approach in a flutter web app with no success,<br>
Currently I have a secret key hardcoded in my app in a *.dart file<br>
Is it possible to access the github secrets directly in the code?<br>
e.g I have</p>
<pre><code class="lang-auto">final secret = "*******";
</code></pre>
<p>can I replace it with</p>
<pre><code class="lang-auto">final secret =${{secrets.myKey }}
</code></pre>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="maheshmnj" data-post="17" data-topic="135875">
<div class="title">
<div class="quote-controls"></div>
 maheshmnj:</div>
<blockquote>
<p>Is it possible to access the github secrets directly in the code?</p>
</blockquote>
</aside>
<p>No, because your code isn’t processed by the workflow parser. In general it’s better to read the secret value from an environment variable or configuration file at runtime. Your workflow can then set that environment variable from a secret, or write the configuration file before calling your application.</p>

**Comment by maheshj01** [comment]
<p>Thanks for the reply<br>
I am not much familiar with these scripts</p>
<pre><code class="lang-auto">- name: execute py script # run sj-gobierno.py to get the latest data
  env:
    EMAIL_ADDRESS: ${{ secrets.EMAIL_ADDRESS }}
    EMAIL_PASSWORD: ${{ secrets.EMAIL_PASSWORD }}
</code></pre>
<p>Does the env file has to be already present or will it get created with the content?</p>
<p>and what should the run  key do I  did not specify the run key and the test  failed with an error</p>
<pre><code class="lang-auto">every step must define a `uses` or `run` key```</code></pre>

**Comment by airtower-luna** [comment]
<p>The <code>run</code> key defines the command to run. Based the comment in your example it should probably look similar to this:</p>
<pre><code class="lang-yaml">- name: execute py script
  run: python3 sj-gobierno.py
  env:
    EMAIL_ADDRESS: ${{ secrets.EMAIL_ADDRESS }}
    EMAIL_PASSWORD: ${{ secrets.EMAIL_PASSWORD }}
</code></pre>
<p>What you define in <code>env</code> will be available to the script as environment variables, not in a file. For accessing environment variables in Python, see <a href="https://docs.python.org/3/library/os.html#os.environ"><code>os.environ</code></a>.</p>

**Comment by ozcanteters** [comment]
Hi from 2023,
how to i use this on c#?
